### PR TITLE
[ORCA-204] Implement Docker submission execution timeout

### PR DESCRIPTION
**Pull request by jaymedina**
### problem

The Docker container submissions run without a time limit, leaving the pipeline vulnerable to hanging executions that could incur unnecessary costs.

### solution

- [x] New params `container_timeout` and `poll_interval` are introduced allowing users to set an upper limit for how long a container should run before it times out, and how long to wait between polling the container for status update
- [x] New function `monitor_container` is introduced for monitoring the elapsed time during a container execution, and shutting it down when necessary
- [x] A new error message `timeout_msg` is generated from `monitor_container` if a container times out, and is propagated thru the pipeline so it reaches the user
- [x] Documentation is updated reflecting new params

### testing & preview

- [x] [Test 1](https://tower-dev.sagebionetworks.org/orgs/Sage-Bionetworks/workspaces/example-dev-project/watch/3Nig3AuizU48n2): Container runs for 2 mins, but `container_timeout` is 1 min

image

- [x] [Test 2](https://tower-dev.sagebionetworks.org/orgs/Sage-Bionetworks/workspaces/example-dev-project/watch/2RAltzyiXYLOj5): Container runs for 2 mins, but `container_timeout` is 5 min

image

- [x] [Test 3](https://tower-dev.sagebionetworks.org/orgs/Sage-Bionetworks/workspaces/example-dev-project/watch/1CmScTlafYAqDG): Testing with multiple submissions of various execution times (2, 4, 5, 10 minutes), but `container_timeout` is 5 min. Ensure that the 1 and 4 minute submissions make it thru.

The expected evaluations (-55 and -56) failed

image

- [x] [Test 4](https://tower-dev.sagebionetworks.org/orgs/Sage-Bionetworks/workspaces/example-dev-project/watch/1iqzPmtQtWjQ32): Multiple submissions (2, 2, 4, 5, 5, 5, 10 minutes), but `container_timeout` is 4 minutes. Only 2 submissions should pass.

As expected, only 9744479 and 9744460 pass:

image

Docker log example for a failed execution:

image

Annotations:

image

----
Remaining questions/comments:
* Users should be notified that there is a max to the amt of time their container can run for. this should be in the participant documentation somewhere. notify verena & gaia
* It would be unfortunate to make a submission, wait X hours, and then get a notification from the organizer that the submission never went thru. how can we notify users that their submissions are taking suspiciously long? is that our responsibility to begin with?

### CODE REVIEW

File: bin/run_docker.py [hunk 1]
```diff
@@ -267,8 +268,55 @@ def mount_volumes() -> dict:
...
+        container: The Docker container subject to monitor.
+        timeout: Maximum duration to monitor the container (in minutes) before forced shutdown.
+        poll_interval: Time to wait between status checks (in minutes).
+        elapsed_time: The elapsed time already spent monitoring (in minutes).
+    """
+
+    # Check if the elapsed time has reached or exceeded the timeout
+    if elapsed_time >= timeout:
+        print("Timeout reached. Stopping container.")
+        container.kill()
```
Comment (thomasyu888) [inline bin/run_docker.py]: Stop then kill.  Look at the existing challengeWorkflowtemplates code here for an example

Comment (jaymedina) [reply bin/run_docker.py]: Awesome, that seems to have done the trick, thank you! I used `stop(timeout=10)` to allow the container enough time to properly close out and this seems to have worked. I'll be running one more test just to be sure it doesn't give an issue, and I think it's a good idea to put this ticket thru a well-vetted validation phase just in case.

File: bin/run_docker.py [hunk 2]
```diff
@@ -267,8 +268,55 @@ def mount_volumes() -> dict:
...
+
+    Arguments:
+        container: The Docker container subject to monitor.
+        timeout: Maximum duration to monitor the container (in minutes) before forced shutdown.
+        poll_interval: Time to wait between status checks (in minutes).
+        elapsed_time: The elapsed time already spent monitoring (in minutes).
+    """
+
+    # Check if the elapsed time has reached or exceeded the timeout
+    if elapsed_time >= timeout:
```
Comment (BWMac) [inline bin/run_docker.py]: Does this always give each submission exactly the same maximum amount of time to execute?

**Review by BWMac — APPROVED** [review]
LGTM! just one question

File: bin/run_docker.py [hunk 3]
```diff
@@ -267,8 +268,55 @@ def mount_volumes() -> dict:
...
+        print("Timeout reached. Stopping container.")
+        container.stop(timeout=10)
+        return f"Container exceeded execution time limit of {timeout} minutes. Unable to process."
+
+    # Refresh container status
+    container.reload()
+
+    # If the container has exited, stop monitoring
+    if container.status == "exited":
```
Comment (thomasyu888) [inline bin/run_docker.py:384]: I would take a look at this logic a bit.  There is a scenario where the container finishes at the last cycle, but then it it'll be  "elapsed_time >= timeout" causing the error to potentially be raised.

File: bin/run_docker.py [hunk 4]
```diff
@@ -267,8 +268,55 @@ def mount_volumes() -> dict:
...
+    # Increment the elapsed time
+    elapsed_time += poll_interval
+
+    # Notify the backend when elapsed time is 3/4 of the way to timeout
+    if elapsed_time >= timeout * 0.75:
+        print(f"Time spent monitoring container (minutes): {elapsed_time}")
+        print(f"Container run will shut down after {timeout} minute(s).")
+
+    # Recursively call the function to continue monitoring
+    return monitor_container(container, timeout, poll_interval, elapsed_time)
```
Comment (thomasyu888) [inline bin/run_docker.py]: Nit: this could be a `while` loop, but the recursion is unavoidable.  Why you would want to do a  `while` loop is so that people can read the code directly and automatically understand the conditions for the code to keep running.

File: bin/run_docker.py (hunk 2 above)
Comment (BryanFauble) [reply bin/run_docker.py]: Maximum time would probably end up being `container_timeout` + `poll_interval` - assume a situation where polling executes a second before timeout. The container would get an extra 10 minutes, but would eventually time out.

File: bin/run_docker.py (hunk 3 above)
Comment (BryanFauble) [reply bin/run_docker.py:384]: Yeah. If I understand, `container.reload()` should probably be the first thing this does. Thinking:

1. reload
2. Check if status `exited`
3. Check if exceeding timeout

Comment (jaymedina) [reply bin/run_docker.py:384]: Nice, thank you both for considering this edge case. Bryan, I can confirm this change works and pushed it in [this commit](https://github.com/Sage-Bionetworks-Workflows/nf-synapse-challenge/pull/26/commits/a152d13797d1a95a7a1a99da64da1731bb92c788). Below are some test results for the case where my docker model runs for 1.5 minutes, and the timeout limit is 2 mins, but the polling interval is 1 min:

before change: `check time -> reload -> check status`

`run_docker.py` thinks the container exceeded the 2min time limit and the container is stopped

Screenshot 2024-05-16 at 11 33 11 AM

<br>
<br>

after change: `reload -> check status -> check time`

`run_docker.py` checks for status before checking for time, preventing a failure for the `timeout - poll_interval < docker run time < timeout` edge case

Screenshot 2024-05-16 at 11 30 29 AM

**Review by BryanFauble — APPROVED** [review]
:fire: LGTM!

File: bin/run_docker.py (hunk 4 above)
Comment (jaymedina) [reply bin/run_docker.py]: Updated: https://github.com/Sage-Bionetworks-Workflows/nf-synapse-challenge/pull/26/commits/802c3f478ff8a3adf1d7082ea2b73a4954581d44

Comment (jaymedina) [reply bin/run_docker.py]: Update: My interpretation of this comment led to adding a `while` loop with an explicit conditional (`while container.status != "exited" or elapsed_time < timeout`) instead of using a `while True`. However, implementing the `while` loop this way while also addressing SonarCloud's requirements complicates the code more than making it easier to read, in my opinion. So what I did instead is turn the while loop into an if statement (https://github.com/Sage-Bionetworks-Workflows/nf-synapse-challenge/pull/26/commits/fa2ce5d3bed9bba3a2da673c269e429b100b524b), and that serves the purpose of giving a reader an at-a-glance view of the conditions that need to be met for the recursion to continue.

let me know if this also works for you

File: bin/run_docker.py (hunk 2 above)
Comment (jaymedina) [reply bin/run_docker.py]: Yes, at most the amount of time the container can run will be up to `timeout` + `poll_interval` - 1 (unit: minutes)

----
e.g.
* A docker container runs for 14 minutes 59 seconds
* the timeout limit is 11 mins
* the poll interval is 5 mins

At min=0
* container status = `"running"`
* `elapsed_time` = 0, AKA under `timeout`, so it keeps running
* `elapsed_time` becomes `elapsed_time` + 5 min
* the monitor sleeps for 5 mins

At min=5
* container status = `"running"`
* `elapsed_time` = 5, AKA under `timeout`, so it keeps running
* `elapsed_time` becomes `elapsed_time` + 5 min
* the monitor sleeps for 5 mins

At min=10
* container status = `"running"`
* `elapsed_time` = 10, AKA under `timeout`, so it keeps running
* `elapsed_time` becomes `elapsed_time` + 5 min
* the monitor sleeps for 5 mins

At min=15
* container status = `"exited"`, the function returns an empty string
----

So a docker submission that extends beyond the allotted time limit can definitely sneak through without a timeout error, which is not ideal, but based on the scenarios I've explored, all cases beyond `timeout` + `poll_interval` - 1 will fail. Enforcing the timeout while also accommodating for `timeout - poll_interval < docker run time < timeout` edge cases is something I can explore in a future ticket.

**Comment by jaymedina** [comment]
[I reran Test 4](https://tower-dev.sagebionetworks.org/orgs/Sage-Bionetworks/workspaces/example-dev-project/watch/5dTFhrhijBKLOM) to ensure things are still working as expected. Passed ✅ 

image

File: bin/run_docker.py (hunk 2 above)
Comment (BryanFauble) [reply bin/run_docker.py]: I think the more immediate (and simple) way is to do something like a min function where the `time.sleep` will sleep either the `poll_interval` or the remaining time to reach `timeout` if `poll_interval` + elapsed time is > `timeout`

Comment (jaymedina) [reply bin/run_docker.py]: I was actually thinking of removing `poll_interval` as a user-fed parameter and calculating it programmatically based on the `timeout` that's fed. I'd like to give this some more thought before pushing anything into `main` related to this. The default `timeout` and `poll_interval` work for PEGS since the `poll_interval` is much smaller than the `timeout`.

Comment (jaymedina) [reply bin/run_docker.py]: Thanks @BryanFauble - I originally thought to remove the `poll_interval` as an argument and compute it behind the scenes to avoid things like overlap in the final poll interval, but I think it's best to leave it exposed as a parameter in case we need to modify it for troubleshooting or the like. I agree your suggestion is the most straight-forward way to tackle this and I've implemented it in https://github.com/Sage-Bionetworks-Workflows/nf-synapse-challenge/pull/26/commits/d040b746237a3ac2bc83fa503232567c50fee263.

Here are some test results from running the script locally:

test 1) `poll_interval` = 3 mins (longer than the `timeout`)
```
(synapseclient) (synapseclient) ip-10-0-0-173:nf-synapse-challenge jmedina$ python bin/run_docker.py 9744669 2 3 50 

UPGRADE AVAILABLE

A more recent version of the Synapse Client (4.2.0) is available. Your version (4.1.1) can be upgraded by typing:
    pip install --upgrade synapseclient

Python Synapse Client version 4.2.0 release notes

https://python-docs.synapse.org/news/

Running container... docker.synapse.org/syn51317219/runs_1d5min@sha256:1136a32f2bfbeb38be8667413cbde052cc88f063e512a519d66051592854a399
Container still running. Checking again in 2.0 minutes.
Time spent monitoring container (minutes): 2.0
Container run will shut down after 2.0 minute(s).
Original log message exceeds 50 Kb. Truncating...
Truncated log file created: 9744669_docker.log
File renamed successfully.
Original file:  /Users/jmedina/Documents/forks/nf-synapse-challenge/output/predictions.csv
Renamed file:  /Users/jmedina/Documents/forks/nf-synapse-challenge/output/9744669_predictions.csv
```

test 2) `poll_interval` = 0.5 mins
```
(synapseclient) (synapseclient) ip-10-0-0-173:nf-synapse-challenge jmedina$ python bin/run_docker.py 9744669 2 0.5 50 

UPGRADE AVAILABLE

A more recent version of the Synapse Client (4.2.0) is available. Your version (4.1.1) can be upgraded by typing:
    pip install --upgrade synapseclient

Python Synapse Client version 4.2.0 release notes

https://python-docs.synapse.org/news/

Running container... docker.synapse.org/syn51317219/runs_1d5min@sha256:1136a32f2bfbeb38be8667413cbde052cc88f063e512a519d66051592854a399
Container still running. Checking again in 0.5 minutes.
Container still running. Checking again in 0.5 minutes.
Container still running. Checking again in 0.5 minutes.
Time spent monitoring container (minutes): 1.5
Container run will shut down after 2.0 minute(s).
Container still running. Checking again in 0.5 minutes.
Time spent monitoring container (minutes): 2.0
Container run will shut down after 2.0 minute(s).
Original log message exceeds 50 Kb. Truncating...
Truncated log file created: 9744669_docker.log
File renamed successfully.
Original file:  /Users/jmedina/Documents/forks/nf-synapse-challenge/output/predictions.csv
Renamed file:  /Users/jmedina/Documents/forks/nf-synapse-challenge/output/9744669_predictions.csv
```

test 3) `poll_interval` = 1.5 mins, `timeout` = 2 mins, which means the last poll would exceed `timeout`
```
(synapseclient) (synapseclient) ip-10-0-0-173:nf-synapse-challenge jmedina$ python bin/run_docker.py 9744669 2 1.5 50 

UPGRADE AVAILABLE

A more recent version of the Synapse Client (4.2.0) is available. Your version (4.1.1) can be upgraded by typing:
    pip install --upgrade synapseclient

Python Synapse Client version 4.2.0 release notes

https://python-docs.synapse.org/news/

Running container... docker.synapse.org/syn51317219/runs_1d5min@sha256:1136a32f2bfbeb38be8667413cbde052cc88f063e512a519d66051592854a399
Container still running. Checking again in 1.5 minutes.
Time spent monitoring container (minutes): 1.5
Container run will shut down after 2.0 minute(s).
Container still running. Checking again in 0.5 minutes.
Time spent monitoring container (minutes): 2.0
Container run will shut down after 2.0 minute(s).
Original log message exceeds 50 Kb. Truncating...
Truncated log file created: 9744669_docker.log
File renamed successfully.
Original file:  /Users/jmedina/Documents/forks/nf-synapse-challenge/output/predictions.csv
Renamed file:  /Users/jmedina/Documents/forks/nf-synapse-challenge/output/9744669_predictions.csv
```

**Comment by jaymedina** [comment]
New tower run with the latest changes: https://tower-dev.sagebionetworks.org/orgs/Sage-Bionetworks/workspaces/example-dev-project/watch/4KbstwkatkxsaK

Once the workflow runs I'll be merging this in
